java/org/apache/tomcat/util/net/openssl/OpenSSLEngine.java

```diff
@@ -141,7 +142,7 @@
     private boolean handshakeFinished;
     private int currentHandshake;
     private boolean receivedShutdown;
-    private volatile boolean destroyed;
+    private final AtomicBoolean destroyed = new AtomicBoolean(false);
```

Comment:
OpenSSLContext has a very similar protection (this is were I got the idea for this solution), but it uses AtomicInteger instead of AtomicBoolean (not sure why, since AtomicBoolean is more semantically "readable", and both implementations internally use a volatile int for storage, so memory usage is the same).

Making both "networkBIO" and "ssl" final would not hurt, but it would likely not resolve the issue either. The problem arises if an exception is thrown within the constructor, and the JVM concurrently invokes finalize() on the same condition. Even if the variables are final, under a failed constructor scenario, they would not be initialized (so both would be zero while the concurrent finalize() executes)